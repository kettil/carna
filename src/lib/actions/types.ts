import { Watermark, Watermarks } from 'istanbul-lib-report';
import { SpawnKillHandler } from '../utils/createSpawnKillHandler';

type BabelActionProps = {
  watch?: boolean;
  spawnKillHandler?: SpawnKillHandler;
};

type BabelNodeActionProps = {
  watch: boolean;
  script: string;
  watchPaths?: string[];
};

type CommitlintActionProps = {
  edit: string;
};

type CoverageWatermarkThreshold = Watermark | number | undefined;
type CoverageActionProps = {
  projects: string[];
  watermarks?: Partial<Record<keyof Watermarks, CoverageWatermarkThreshold>>;
};

type DepcheckActionProps = {
  path: string;
  ignorePackages?: string[];
};

type EslintActionProps = {
  write?: boolean;
  files?: string[];
};

type JestActionProps = {
  projects: string[];
  updateSnapshot?: boolean;
  runInBand?: boolean;
  verbose?: boolean;
  watch?: boolean;
};

type LicensecheckActionProps = {
  ignorePackages: string[];
  path?: string;
};

type PrettierActionProps = {
  write?: boolean;
  extension?: string;
  files?: string[];
};

type SemverActionProps = {
  path?: string;
};

type TscActionProps = {
  mode: 'type-check' | 'type-create';
};

export type {
  BabelActionProps,
  BabelNodeActionProps,
  CommitlintActionProps,
  CoverageActionProps,
  CoverageWatermarkThreshold,
  DepcheckActionProps,
  EslintActionProps,
  JestActionProps,
  LicensecheckActionProps,
  PrettierActionProps,
  SemverActionProps,
  TscActionProps,
};
