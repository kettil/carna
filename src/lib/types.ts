import { Logger } from './cli/logger';

export type NpmInstallMode = 'prod' | 'dev' | 'optional';

export type PropsGlobal = {
  readonly cwd: string;
  readonly tpl: string;
  readonly cfg: string;
  readonly log: Logger;
  readonly vvv: boolean;
};

export type Action<
  Props extends Record<string, unknown> | undefined = undefined,
  ActionReturn = void
> = Props extends undefined
  ? (argv: PropsGlobal) => Promise<ActionReturn>
  : (argv: PropsGlobal, props: Props) => Promise<ActionReturn>;
