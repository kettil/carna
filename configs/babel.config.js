module.exports = {
  ignore: ['**/*.test.js', '**/*.test.ts', '**/*.test.tsx', '**/*.test-d.ts'],
  presets: ['@babel/env', '@babel/preset-typescript'],
  plugins: [
    [
      '@babel/plugin-transform-runtime', // replace helper code with runtime imports (deduplication)
      { corejs: 3 }, // import corejs polyfills exactly where they are needed
    ],
  ],
  env: {
    test: {
      // extra configuration for process.env.NODE_ENV === 'test'
      presets: ['@babel/env'], // overwrite env-config from above with transpiled module syntax
    },
  },
};
