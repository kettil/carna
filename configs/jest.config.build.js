/* eslint-disable import/no-unresolved, node/no-unpublished-require, node/no-missing-require */
module.exports = Object.assign(require('../../../jest.config.js'), {
  coverageReporters: ['text'],
});
