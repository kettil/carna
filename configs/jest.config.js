const commons = {
  bail: 10,

  testEnvironment: 'node',

  clearMocks: true,
  restoreMocks: true,
};

const createProject = (folder, color, config) => ({
  ...commons,

  displayName: { name: folder, color },
  roots: ['<rootDir>/src', `<rootDir>/tests/${folder}`],

  testMatch: ['**/*.test.{js,ts,tsx}'],
  testPathIgnorePatterns: ['/node_modules/', '/src/'],

  ...config,
});

const config = {
  ...commons,
  projects: [],

  collectCoverage: true,
  collectCoverageFrom: [
    '<rootDir>/src/**/*.{ts,tsx,js}',
    '!<rootDir>/src/**/types.ts',
    '!<rootDir>/src/index.ts',

    '<rootDir>/packages/*/src/**/*.{ts,tsx,js}',
    '!<rootDir>/packages/*/src/**/types.ts',
    '!<rootDir>/packages/*/src/index.ts',
  ],
  coveragePathIgnorePatterns: ['/__jest__/'],
  coverageReporters: ['text-summary', 'html'],
  coverageDirectory: 'coverage',
  coverageThreshold: {
    global: {
      statements: 90,
      branches: 90,
      functions: 90,
      lines: 90,
    },
  },
};

module.exports = { config, createProject };
