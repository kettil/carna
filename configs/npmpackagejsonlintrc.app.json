{
  "extends": "@kettil/npm-package-json-lint-config/app"
}
