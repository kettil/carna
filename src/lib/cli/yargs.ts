import { CommandModule, Arguments, Argv } from 'yargs';
import ExecutableError from '../errors/executableError';
import { PropsGlobal } from '../types';

export type CommandModuleCommand = string;

export type CommandModuleDescribe = NonNullable<CommandModule<PropsGlobal>['describe']>;

export type CommandModuleBuilder<Props = Record<string, unknown>> = (
  yargs: Argv<PropsGlobal>,
) => Argv<PropsGlobal & Props>;

export type CommandModuleHandler<Props = Record<string, unknown>> = (args: Arguments<PropsGlobal & Props>) => void;

export const builderDefault = <Props>(cmd: string, fn: CommandModuleBuilder<Props>): CommandModuleBuilder<Props> => (
  yargs,
) => fn(yargs).usage(`Usage: $0 ${cmd} [options]`).help().version(false)
  .showHelpOnFail(false);

export const errorHandler = (argv: PropsGlobal, error: unknown): void => {
  if (error instanceof ExecutableError) {
    argv.log.log(error.stdout);

    /* eslint-disable-next-line node/no-process-exit, unicorn/no-process-exit */
    process.exit(1);
  }

  throw error;
};
