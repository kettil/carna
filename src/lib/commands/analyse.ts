import { createBuilder, createHandler } from '../cli/yargs';
import type { AnalyseProps } from '../tasks/analyseTask';
import { analyseServices, analyseTask } from '../tasks/analyseTask';

const command = 'analyse';
const desc = 'Run code quality tools';

const aliases = ['$0'];
const options = { group: `${command}-Options` } as const;
const boolOptions = { ...options, type: 'boolean', default: false } as const;

const handler = createHandler<AnalyseProps>(analyseTask);
const builder = createBuilder<AnalyseProps>(command, (yargs) =>
  yargs.options({
    all: {
      ...boolOptions,
      alias: 'a',
      describe: 'All files are checked and not only the changed/new files',
    },
    only: {
      ...options,
      alias: 'o',
      choices: analyseServices,
      describe: 'Run a single code quality tool',
    },
  }),
);

export { aliases, command, desc, builder, handler };
