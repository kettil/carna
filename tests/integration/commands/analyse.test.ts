import * as fs from 'fs';
import { Argv } from 'yargs';
import { handler, builder, command, desc } from '../../../src/lib/commands/analyse';
import { PropsGlobal } from '../../../src/lib/types';
import { getAccessFiles, getArgv, getReadFileWithHooksFiles } from '../../shared/configs';

describe('command analyse', () => {
  test('it should be complete the yargs command structure', () => {
    expect(typeof command).toBe('string');
    expect(typeof desc).toBe('string');
    expect(typeof builder).toBe('function');
    expect(typeof handler).toBe('function');
  });

  test('it should be create the expected builder', () => {
    const yargs = {
      options: jest.fn((...args: unknown[]) => {
        expect(args).toMatchSnapshot('options');

        return yargs;
      }),
      usage: jest.fn((...args: unknown[]) => {
        expect(args).toMatchSnapshot('usage');

        return yargs;
      }),

      help: jest.fn(() => yargs),
      version: jest.fn(() => yargs),
      showHelpOnFail: jest.fn(() => yargs),
    } as unknown as Argv<PropsGlobal>;

    builder(yargs);
  });

  test('it should work', async () => {
    const result = await handler(getArgv({}));

    expect(result).toBeUndefined();
  });

  test('it should work with hooks', async () => {
    (fs as any).setMockReadFileFiles(getReadFileWithHooksFiles());

    const result = await handler(getArgv({}));

    expect(result).toBeUndefined();
  });

  test.each<['eslint' | 'prettier' | 'typescript']>([['eslint'], ['prettier'], ['typescript']])(
    'it should work with only %p',
    async (only) => {
      (fs as any).setMockAccessFiles(getAccessFiles());

      const result = await handler(getArgv({ only }));

      expect(result).toBeUndefined();
    },
  );
});
