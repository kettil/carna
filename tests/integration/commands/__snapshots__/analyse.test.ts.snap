// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`command analyse it should be create the expected builder: options 1`] = `
Array [
  Object {
    "all": Object {
      "alias": "a",
      "default": false,
      "describe": "All files are checked and not only the changed/new files",
      "group": "analyse-Options",
      "type": "boolean",
    },
    "only": Object {
      "alias": "o",
      "choices": Array [
        "eslint",
        "prettier",
        "typescript",
      ],
      "describe": "Run a single code quality tool",
      "group": "analyse-Options",
    },
  },
]
`;

exports[`command analyse it should be create the expected builder: usage 1`] = `
Array [
  "Usage: $0 analyse [options]",
]
`;

exports[`command analyse it should work with hooks: child-process 1`] = `
Object {
  "args": Array [
    "run",
    "precarna:analyse",
  ],
  "command": "npm",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command analyse it should work with hooks: child-process 2`] = `
Object {
  "args": Array [
    "--config",
    "/path/to/project/node_modules/carna/configs/prettierrc.json",
    "--write",
    "\\"./**/*.{ts,tsx,js,jsx,json,md,css,scss,yml,yaml,html}\\"",
  ],
  "command": "/path/to/project/node_modules/.bin/prettier",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command analyse it should work with hooks: child-process 3`] = `
Object {
  "args": Array [
    "--color",
    "--max-warnings",
    "0",
    "--config",
    "/path/to/project/node_modules/carna/configs/eslintrc.json",
    "--fix",
    "--ext",
    "\\"js,ts,tsx\\"",
    ".",
  ],
  "command": "/path/to/project/node_modules/.bin/eslint",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command analyse it should work with hooks: child-process 4`] = `
Object {
  "args": Array [
    "--outDir",
    "/path/to/project/build",
    "--project",
    "/path/to/project/tsconfig.json",
    "--noEmit",
    "--isolatedModules",
    "false",
  ],
  "command": "/path/to/project/node_modules/.bin/tsc",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command analyse it should work with hooks: child-process 5`] = `
Object {
  "args": Array [
    "run",
    "postcarna:analyse",
  ],
  "command": "npm",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command analyse it should work with only "eslint": child-process 1`] = `
Object {
  "args": Array [
    "--color",
    "--max-warnings",
    "0",
    "--config",
    "/path/to/project/node_modules/carna/configs/eslintrc.json",
    "--fix",
    "--ext",
    "\\"js,ts,tsx\\"",
    ".",
  ],
  "command": "/path/to/project/node_modules/.bin/eslint",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command analyse it should work with only "prettier": child-process 1`] = `
Object {
  "args": Array [
    "--config",
    "/path/to/project/node_modules/carna/configs/prettierrc.json",
    "--write",
    "\\"./**/*.{ts,tsx,js,jsx,json,md,css,scss,yml,yaml,html}\\"",
  ],
  "command": "/path/to/project/node_modules/.bin/prettier",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command analyse it should work with only "typescript": child-process 1`] = `
Object {
  "args": Array [
    "--outDir",
    "/path/to/project/build",
    "--project",
    "/path/to/project/tsconfig.json",
    "--noEmit",
    "--isolatedModules",
    "false",
  ],
  "command": "/path/to/project/node_modules/.bin/tsc",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command analyse it should work: child-process 1`] = `
Object {
  "args": Array [
    "--config",
    "/path/to/project/node_modules/carna/configs/prettierrc.json",
    "--write",
    "\\"./**/*.{ts,tsx,js,jsx,json,md,css,scss,yml,yaml,html}\\"",
  ],
  "command": "/path/to/project/node_modules/.bin/prettier",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command analyse it should work: child-process 2`] = `
Object {
  "args": Array [
    "--color",
    "--max-warnings",
    "0",
    "--config",
    "/path/to/project/node_modules/carna/configs/eslintrc.json",
    "--fix",
    "--ext",
    "\\"js,ts,tsx\\"",
    ".",
  ],
  "command": "/path/to/project/node_modules/.bin/eslint",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;

exports[`command analyse it should work: child-process 3`] = `
Object {
  "args": Array [
    "--outDir",
    "/path/to/project/build",
    "--project",
    "/path/to/project/tsconfig.json",
    "--noEmit",
    "--isolatedModules",
    "false",
  ],
  "command": "/path/to/project/node_modules/.bin/tsc",
  "cwd": "/path/to/project",
  "env": Object {},
}
`;
